text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+    fn my_default() -> Self;
+}
+```
+
+Right now, with what we know, we have to copy-paste our impl for every integer type. For example, an impl for `u8` would be:
+
+```rust
+impl MyDefault for u8 {
+    fn my_default() -> u8 { 1 }
+}
+```
+
+Except, we'd have to replicate this for every single integer type. If we're clever, we could use a macro:
+
+```rust
+macro_rules! impl_my_default {
+    ($($i: ident),*) => {$(
+        impl MyDefault for $i {
+            fn my_default() -> $i { 1 }
+        }
+    )*}
+}
+impl_my_default!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
+```
+
+But really, that's just making the compiler do the copy-pasting for us. Instead, we're going to use the special types `u<N>` and `i<N>` to generalise the code for us. The end result looks like:
+
+```rust
+impl<const N: usize> MyDefault for u<N> {
+    fn my_default() -> u<N> { 1 }
+}
+impl<const N: usize> MyDefault for i<N> {
+    fn my_default() -> i<N> { 1 }
+}
+impl MyDefault for usize {
+    fn my_default() -> usize { 1 }
+}
+impl MyDefault for isize {
+    fn my_default() -> isize { 1 }
+}
+```
+
+That's a lot better. Now, as you'll notice, we still have to cover the types `usize` and `isize` separately; that's because they're still separate from the `u<N>` and `i<N>` types. If you think about it, this has always been the case before generic integers; for example, on a 64-bit system, `u64` is not the same as `usize`.
+
+## Zero-sized integers
+
+There's one slight caveat here: our `my_default` method might overflow. This seems silly, but there's three types, `i<1>`, `u<0>`, and `i<0>`, which can't have the value 1. In general, if you're casting a literal *to* a generic integer, you can't expect any value other than zero to work. In the future, we'll be able to annotate our `i<N>` impl with something like `where N > 1` or `where N >= 8`, but until then, we'll have to deal with this weird overflow behaviour.
+
+The rules you'd expect apply to `u<N>` and `i<N>`, which is that `u<N>` stores values in the range `0..2.pow(N)`, and `i<N>` stores integers in the range `-2.pow(N - 1)..2.pow(N - 1)`. This means that `u<1>` only holds the values `0` and `1`, and `i<1>` only holds the values `-1` and `0`. The meaning for `u<0>` and `i<0>` is a little less clear, but they both are only allowed to contain the value `0`; the ranges end up being `0..1` and `-0.5..0.5`, which… yeah, zero is the only integer in those ranges, but it still can be confusing.
+
+For now, if you want to ensure that your integers are the right size, you can add a `const { ... }` assertion to your implementations like so:
+
+```rust
+impl<const N: usize> MyDefault for u<N> {
+    fn my_default -> u<N> {
+        const { assert!(N > 1); }
+        1
+    }
+}
+```
+
+This will cause the compiler to fail when `MyDefault` is used for `u<0>` or `u<1>`, since it will force the constant block to be evaluated. Not ideal, but it's the best we've got for now.
+
+## Uncommonly sized integers
```

Comment:
Here are two answers. Their order was chosen by an unbiased coin flip.

Obviously, the value should be negative one, because two's complement always has one more negative number than positive numbers.

Obviously, the value should be zero, because `i<N>` and `u<N>` always have this value in common.

Reply:
Less tongue-in-cheek discussion about why/whether i0 should be permitted (in certain contexts) and what its value should be: https://discourse.llvm.org/t/rfc-support-zero-width-integers-in-the-comb-dialect-circt/78492

Reply:
Or `i<0>` could be an uninhabitable type.

Reply:
If it was uninhabited, every possible way of constructing an `i<N>` needs to special case N = 0, either with bounds that rule it out (not properly supported yet, and likely implies an annotation burden) or by panicking. That sounds even more inconvenient than a post-mono error.

Reply:
Not sure if you just missed this part, but I explicitly clarified the value of `i<0>` with this:

> `u<N>` are able to store integers in the range `0..2.pow(N)` and `i<N>` are able to store integers in the range `-2.pow(N-1)..2.pow(N-1)`. The cheeky specificity of "integers in the range" ensures that, for `i<0>`, the range `-0.5..0.5` only contains the integer zero; in general, `u<0>` and `i<0>` will need to be special-cased anyway, as they must be ZSTs.

This felt like the most logical approach.

Reply:
Also a side note on uninhabited types: I mention later that it's an unanswered question whether `NonZero<u<0>>` would be allowed and uninhabited, but my assumption would be yes. So, it makes sense that the weird special-casing of zero would be reserved to the types that explicitly call out zero as a special case.